homeassistant/components/hdmi_cec.py

```diff
@@ -4,118 +4,535 @@
-        parents = []
-    for addr, val in mapping.items():
-        cur = parents + [str(addr)]
-        if isinstance(val, dict):
-            yield from parse_mapping(val, cur)
-        elif isinstance(val, str):
-            yield (val, cur)
+SERVICE_SEND_COMMAND_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(ATTR_CMD): vol.Coerce(int),
+        vol.Optional(ATTR_SRC): vol.Coerce(int),
+        vol.Optional(ATTR_DST): vol.Coerce(int),
+        vol.Optional(ATTR_ATT): vol.Coerce(int),
+        vol.Optional(ATTR_RAW): vol.Coerce(str)
+    })
+}, extra=vol.REMOVE_EXTRA)
 
+SERVICE_VOLUME_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(CMD_UP): vol.Coerce(int),
+        vol.Optional(CMD_DOWN): vol.Coerce(int),
+        vol.Optional(CMD_MUTE): None,
+        vol.Optional(CMD_UNMUTE): None,
+        vol.Optional(CMD_MUTE_TOGGLE): None
+    })
+}, extra=vol.REMOVE_EXTRA)
 
-def pad_physical_address(addr):
-    """Right-pad a physical address."""
-    return addr + ['0'] * (MAX_DEPTH - len(addr))
+SERVICE_POWER_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(ATTR_DST): vol.Coerce(int),
+    })
+}, extra=vol.REMOVE_EXTRA)
 
 
-def setup(hass, config):
+def setup(hass, base_config):
     """Setup CEC capability."""
-    global _CEC
-
-    try:
-        import cec
-    except ImportError:
-        _LOGGER.error("libcec must be installed")
-        return False
-
-    # Parse configuration into a dict of device name to physical address
-    # represented as a list of four elements.
-    flat = {}
-    for pair in parse_mapping(config[DOMAIN].get(CONF_DEVICES, {})):
-        flat[pair[0]] = pad_physical_address(pair[1])
-
-    # Configure libcec.
-    cfg = cec.libcec_configuration()
-    cfg.strDeviceName = 'HASS'
-    cfg.bActivateSource = 0
-    cfg.bMonitorOnly = 1
-    cfg.clientVersion = cec.LIBCEC_VERSION_CURRENT
-
-    # Setup CEC adapter.
-    _CEC = cec.ICECAdapter.Create(cfg)
-
-    def _power_on(call):
-        """Power on all devices."""
-        _CEC.PowerOnDevices()
-
-    def _standby(call):
-        """Standby all devices."""
-        _CEC.StandbyDevices()
-
-    def _select_device(call):
-        """Select the active device."""
-        path = flat.get(call.data[ATTR_DEVICE])
-        if not path:
-            _LOGGER.error("Device not found: %s", call.data[ATTR_DEVICE])
-        cmds = []
-        for i in range(1, MAX_DEPTH - 1):
-            addr = pad_physical_address(path[:i])
-            cmds.append('1f:82:{}{}:{}{}'.format(*addr))
-            cmds.append('1f:86:{}{}:{}{}'.format(*addr))
-        for cmd in cmds:
-            _CEC.Transmit(_CEC.CommandFromString(cmd))
-        _LOGGER.info("Selected %s", call.data[ATTR_DEVICE])
+
+    _LOGGER.debug("CEC setup")
+    config = base_config.get(DOMAIN)
+
+    global CEC_CLIENT
+    CEC_CLIENT = CecClient(hass)
+    exclude = config.get(CONF_EXCLUDE)
+    stop = threading.Event()
 
     def _start_cec(event):
```

Comment:
You cannot have an event handler that never returns, you are then stealing a thread from HASS. If you want to run in a thread, start your own.